src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1490,16 +1491,30 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
             throw new LogicException(sprintf('The default bus named "%s" is not defined. Define it or change the default bus name.', $config['default_bus']));
         }
 
-        $messageToSenderIdsMapping = array();
+        $messageToSenderIdMapping = array();
+        $messageToSendAndHandleMapping = array();
         foreach ($config['routing'] as $message => $messageConfiguration) {
             if ('*' !== $message && !class_exists($message) && !interface_exists($message, false)) {
                 throw new LogicException(sprintf('Messenger routing configuration contains a mistake: message "%s" does not exist. It needs to match an existing class or interface.', $message));
             }
 
-            $messageToSenderIdsMapping[$message] = $messageConfiguration['senders'];
+            if (1 < count($messageConfiguration['senders'])) {
+                $senders = array_map(function ($sender) { return new Reference($sender); }, $messageConfiguration['senders']);
+                $chainSenderDefinition = new Definition(ChainSender::class, array($senders));
+                $chainSenderId = 'messenger.chain_sender.'.$message;
```

Comment:
You should prefix the service name with a dot, so it's hidden to the user.

Reply:
done